### PR TITLE
[docs] Add missing package flatbuffers-dev to Ubuntu build instructions

### PR DESCRIPTION
**Pull request by bugwelle**
## Description
Add missing package `flatbuffers-dev` to Ubuntu 18.04 build instructions

## Motivation and Context
Tried to build Kodi on my machine: Kubuntu 18.04  
Followed steps provided at https://github.com/xbmc/xbmc/blob/master/docs/README.Ubuntu.md

```sh
git clone https://github.com/xbmc/xbmc Kodi
cd Kodi; mkdir build && cd $_
cmake ../Kodi -DCMAKE_INSTALL_PREFIX=/usr/local
```

Error message:

```
CMake Error at /home/andre/.local/lib/python3.6/site-packages/cmake/data/share/cmake-3.14/Modules/FindPackageHandleStandardArgs.cmake:137 (message):
  Could NOT find FlatBuffers (missing: FLATBUFFERS_FLATC_EXECUTABLE
  FLATBUFFERS_INCLUDE_DIR)
```

`flatbuffers-dev` ist listed here as well: https://github.com/xbmc/xbmc/blob/master/docs/README.Linux.md#3-install-the-required-packages

## How Has This Been Tested?
Installed `flatbuffers-dev` and other recommended packages and build Kodi successfully.

## Screenshots (if appropriate):
None.

## Types of change

- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [x] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project (*Note:* Well, just documentation after all...)
- [x] My change requires a change to the documentation, either Doxygen or wiki  
      This *is* a change to the documentation. Don't know about the wiki, though.
- [x] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change (*Note:* No code change)
- [ ] All new and existing tests passed  (*Note:* No code change)

## Other notes
If you don't take minor pull requests like this, just let me know. :smiley: 
Btw: Your build instructions are great. Works like a charm. Kodi also builds way faster than I expected.  :+1:

### CODE REVIEW

**Comment by DaVukovic** [comment]
I agree. This should be added.

**Comment by bugwelle** [comment]
@DaVukovic Do you even backport build instructions? I'm not yet familiar with Kodi's development and backport policies. :-)

File: docs/README.Ubuntu.md
Comment (Rechi) [inline docs/README.Ubuntu.md]: `flatbuffers-dev` is not only needed for Ubuntu 18.04, it's needed everywhere. It has to be added to the common dependency list a few lines above

Comment (bugwelle) [reply docs/README.Ubuntu.md]: You're right. Didn't test it on Ubuntu 16.04.
I'll update this pr tomorrow. Thanks.

**Comment by DaVukovic** [comment]
>Do you even backport build instructions? 

Yes, why not? If we wouldn't Backport that and someone wants to compile Kodi 18.x in the future it might fail. If you switch branches you also get a different version for the build instructions. So I would say that a Backport wouldn't hurt. Even more...I would say itt necessary. ;)

**Comment by bugwelle** [comment]
Hi,

it seems that Ubuntu's build instructions are missing a step. I tried to build Kodi using a simple docker image to check if it works on a clean machine but two packages (one of which is `flatbuffers-dev`) can't be found:


```docker
FROM ubuntu:bionic

RUN apt-get update && \
    apt-get -y --no-install-recommends install \
        debhelper autoconf automake autopoint gettext \
        autotools-dev cmake curl default-jre doxygen flatbuffers-dev \
        gawk gcc gdc gperf libasound2-dev libass-dev libavahi-client-dev \
        libavahi-common-dev libbluetooth-dev libbluray-dev libbz2-dev \
        libcdio-dev libp8-platform-dev libcrossguid-dev libcurl4-openssl-dev \
        libcwiid-dev libdbus-1-dev libegl1-mesa-dev libenca-dev libflac-dev \
        libfontconfig-dev libfreetype6-dev libfribidi-dev libfstrcmp-dev \
        libgcrypt-dev libgif-dev libgles2-mesa-dev libgl1-mesa-dev \
        libglu1-mesa-dev libgnutls28-dev libgpg-error-dev libiso9660-dev \
        libjpeg-dev liblcms2-dev libltdl-dev liblzo2-dev libmicrohttpd-dev \
        libmysqlclient-dev libnfs-dev libogg-dev libpcre3-dev libplist-dev \
        libpng-dev libpulse-dev libshairplay-dev libsmbclient-dev \
        libsqlite3-dev libssl-dev libtag1-dev libtiff5-dev libtinyxml-dev \
        libtool libudev-dev libva-dev libvdpau-dev libvorbis-dev libxmu-dev \
        libxrandr-dev libxslt1-dev libxt-dev lsb-release python-dev \
        python-pil rapidjson-dev swig unzip uuid-dev yasm zip zlib1g-dev

# Ubuntu 18.04 specific
RUN apt install libcec-dev libfmt-dev liblirc-dev

# [...]
```

When building this docker image, I get following errors:

```
E: Unable to locate package flatbuffers-dev
E: Unable to locate package libshairplay-dev
```

On my system those are found at team xbmc's PPA:

```
$ apt show libshairplay-dev                
Package: libshairplay-dev
Version: 0.9.0.2~bionic
Priority: extra
Section: libdevel
Source: shairplay
Maintainer: wsnipex <wsnipex@a1.net>
Installed-Size: 19,5 kB
Depends: libao-dev, libshairplay0 (= 0.9.0.2~bionic)
Download-Size: 4.000 B
APT-Manual-Installed: yes
APT-Sources: http://ppa.launchpad.net/team-xbmc/ppa/ubuntu bionic/main amd64 Packages
Description: Free portable AirPlay server implementation (development files)
 Free portable AirPlay server implementation similar to ShairPort.
```

Section 3.2 does not state to add another repository. 3.1 does but the instructions say that users can either go to 3.1 or 3.2.

Is there another repository where those can be found? The build instructions for ubuntu should explicitly state where those two packages can be found *or* should tell users to build them themselves.

I'll try to create a few docker images and will check if all build instructions are valid (at least I'll do that for Ubuntu and OpenSUSE) :smiley:

**Comment by bugwelle** [comment]
Forget the last part. The documentation already says:

> Take a note of the missing dependencies and, after a successful step completion, build the missing dependencies manually.

Nevertheless, one package can't be found :)

**Comment by bugwelle** [comment]
Ok, so i tested a few distributions and Kodi's build instructions miss a few things.

### Fedora
Install steps for Fedora 28 work well. But for later versions `java-9-openjdk-headless` can't be found. `java-latest-openjdk-headless` instead works well.

For latest Fedora (not tested for older versions) a library is missing:

```
/usr/bin/ld.gold: error: cannot find -lunistring
collect2: error: ld returned 1 exit status
```

### Ubuntu
For Ubuntu 16.04, 18.04 and 19.04 two packages cannot be found:

```
E: Unable to locate package flatbuffers-dev
E: Unable to locate package libshairplay-dev
```

### openSUSE
Tested Tumbleweed. Leap 15.1 had an issue with opensuse-multimedia-libs (I did change the URL to be Leap 15.1 specific).

Missing packages:

```
Loading repository data...
Reading installed packages...
'libjpeg-devel' not found in package names. Trying capabilities.
'libmysqlclient-devel' not found in package names. Trying capabilities.
'gettext-devel' not found in package names. Trying capabilities.
'java-openjdk' not found in package names. Trying capabilities.
'randrproto-devel' not found in package names. Trying capabilities.
'renderproto-devel' not found in package names. Trying capabilities.
No provider of 'randrproto-devel' found.
No provider of 'renderproto-devel' found.
```

Used Dockerfiles available at https://github.com/bugwelle/kodi-docker-build

**Comment by bugwelle** [comment]
Updated my Dockerfiles and build instructions.

For older versions packages are missing (stated below). This is not an issue as the user is told to build them manually if that happens. Nevertheless, all packages should be found on the latest version of each distribution, right?

### Fedora
I've updated Fedora's build instructions. 
`make` wasn't (?) installed in Docker's `fedora:latest`. Adding it to the list of required packages should do no harm. 

Furthermore I still get the same error as in my previous post. Cannot link `unistring`. Installing `libunistring` does not work as well...

 - works: [Fedora 26](https://github.com/bugwelle/kodi-docker-build/blob/master/Dockerfile.build-fedora-26)
 - [Fedora Latest](https://github.com/bugwelle/kodi-docker-build/blob/master/Dockerfile.build-fedora-latest)

Feedback welcome.

### openSUSE
Build instructions explicitly state that any failing package during install should be build by the user.
Should I change anything?
After all, two packages can't be found on Tumbleweed (the latest version).

```
No provider of 'randrproto-devel' found.
No provider of 'renderproto-devel' found.
```

### Ubuntu
I added a note `and later` as the mentioned packages are required on Ubuntu 19.04 as well.
Also `flatbuffers-dev` was added. It won't be found on any Ubuntu version, though...

#### Ubuntu 16.04
`libp8-platform-dev` and `libshairplay-dev` can't be found. Installing the first using `depends` works.  

May be a stupid question, but how do I build `libshairplay`? Building `flatbuffers` works well.

```
Makefile:1: ../../Makefile.include: No such file or directory
make: *** No rule to make target '../../Makefile.include'.  Stop.
make: Entering directory '/tmp/kodi/tools/depends/target/libshairplay'
make: Leaving directory '/tmp/kodi/tools/depends/target/libshairplay'
```

Also `libcec4-dev` and `libfmt3-dev` are not available. Those are listed explicitly for Ubuntu 16.04. That's not intended, is it?

### Ubuntu 18.04
`libshairplay-dev` can't be found. Same as above.  

### Ubuntu 19.04
`libshairplay-dev` can't be found. Same as above.  
Same issue as with 18.04.

<hr>

That's it from me for today. :smiley:

File: docs/README.Fedora.md
Comment (Rechi) [inline docs/README.Fedora.md]: I think you wanted to write `java-latest-openjdk-headless` once.
I'm not sure if it's a good idea to reverence `java-latest-openjdk-headless` instead of a specific version as it may not work with newer versions.

Comment (bugwelle) [reply docs/README.Fedora.md]: Copy&Paste. And I even checked that twice... Thanks for the heads-up.
That would be `java-11-openjdk-headless`.

**Comment by Rechi** [comment]
> ### openSUSE
> ```
> No provider of 'randrproto-devel' found.
> No provider of 'renderproto-devel' found.
> ```

I think those are replaced by `xorgproto-devel` package for Tumbleweed.

> ### Ubuntu
> 
> Also `flatbuffers-dev` was added. It won't be found on any Ubuntu version, though...

That's provided at ppa:team-xbmc/xbmc-nightly.

> #### Ubuntu 16.04
> 
> `libp8-platform-dev` and `libshairplay-dev` can't be found. Installing the first using `depends` works.

Those are provided at ppa:team-xbmc/xbmc-nightly.

> 
> May be a stupid question, but how do I build `libshairplay`? Building `flatbuffers` works well.
> 
> ```
> Makefile:1: ../../Makefile.include: No such file or directory
> make: *** No rule to make target '../../Makefile.include'.  Stop.
> make: Entering directory '/tmp/kodi/tools/depends/target/libshairplay'
> make: Leaving directory '/tmp/kodi/tools/depends/target/libshairplay'
> ```

Where did you read that it should work?

> Also `libcec4-dev` and `libfmt3-dev` are not available. Those are listed explicitly for Ubuntu 16.04. That's not intended, is it?

Those are provided at ppa:team-xbmc/xbmc-nightly and ppa:team-xbmc/xbmc-ppa-build-depends

> ### Ubuntu 18.04
> 
> `libshairplay-dev` can't be found. Same as above.

That's provided at ppa:team-xbmc/xbmc-nightly.

> ### Ubuntu 19.04
> 
> `libshairplay-dev` can't be found. Same as above.
> Same issue as with 18.04.

That's provided at ppa:team-xbmc/xbmc-nightly.

**Comment by bugwelle** [comment]
> That's provided at ppa:team-xbmc/xbmc-nightly.

Indeed. And I'm using that repository on my local machine as well (see https://github.com/xbmc/xbmc/pull/16524#issuecomment-524536747 ).

But the [build instructions](https://github.com/xbmc/xbmc/blob/master/docs/README.Ubuntu.md#31-get-build-dependencies-automagically) only mention that repository in section "[3.1 Get build dependencies automagically](https://github.com/xbmc/xbmc/blob/master/docs/README.Ubuntu.md#31-get-build-dependencies-automagically)". There's no mention of that repository being required in section 3.2. And the introduction of section "3. Install the required packages" states that *either* can be chosen. Also section 3.2 instructs the user to build missing dependencies themselves.

That's why I thought and still think that the repository should not be mandatory (though it's definitely useful) :smiley: 

> Where did you read that it should work?

I kind of expected it to work after reading "[3.1 Build missing dependencies](https://github.com/xbmc/xbmc/blob/master/docs/README.Linux.md#31-build-missing-dependencies)" in the linux build instructions.  That section also refers to the directory `tools/depends/target`. 

Any help is welcome. :)

**Comment by wsnipex** [comment]
repositories are never mandatory, you can always build dependencies yourself. It's Linux after all.

**Review by wsnipex — APPROVED** [review]

**Comment by bugwelle** [comment]
> repositories are never mandatory, you can always build dependencies yourself. It's Linux after all.

Agree. My question is just: Should we add packages to the list of dependencies that we *know* are not available in the standard repositories?

I'm fine with my changes but could add another note if somebody wants that. :smiley: 

Thank you for approving these changes. If other's are fine with this PR as well, could somebody merge it?

**Comment by Rechi** [comment]
For me reading the following step doesn't sound like it's optional. So once you are at the step installing dependencies you already have the required PPA(s) added.
https://github.com/xbmc/xbmc/blob/c79fb23a7d7b8550f236ecbc9d3f0af4e57365e4/docs/README.Ubuntu.md#L72-L82

**Review by Rechi — APPROVED** [review]

**Comment by bugwelle** [comment]
@Rechi Yes, but that's one of two possible ways to install Kodi's dependencies. 
3.2 is the "automagical" way because we can use Kodi's repositories for it.

3. says that there are two options. I can use either one: `[...] one of two methods:`

https://github.com/xbmc/xbmc/blob/c79fb23a7d7b8550f236ecbc9d3f0af4e57365e4/docs/README.Ubuntu.md#L68

If it's unclear then we may want to change the wording of that sentence. 😃  
Thank you for taking the time to review this.

**Comment by Rechi** [comment]
I've read it like the two options are `sudo apt build-dep kodi` or `sudo apt install ...`.
It would be great to change something that `sudo add-apt-repository -s ...` is needed in both cases. Maybe move that block above the 3.1 section.

**Comment by bugwelle** [comment]
Hi @Rechi ,

from my point of view that shouldn't be mandatory but I'm also not the one that wrote the guide so I may be mistaken.  
If you're fine with it, then I'd keep it as it is for now.  :-)

This PR has two approvals, thank you. I can't merge it myself, though. :smiley:

**Pull request merged**
